spmd/compiler/distribute.py

```diff
@@ -421,25 +422,68 @@ def __init__(
         # used to propagate sharding from the output of the forward pass to
         # the input of backward pass
         self._known_specs_by_node_name: Dict[str, Schema] = {}
+        # A switch that allow users to turn off map_param_and_grad since it is
+        # brittle as for now (the impl depends on the param and grad order).
+        self._map_param_and_grad = map_param_and_grad
+
+    def _is_param(self, t: torch.Tensor) -> bool:
+        # N.B.: id(t) and id(param) does not match
+        orig_module = cast(nn.Module, self._dist_graph.orig_module)
+        return t.storage().data_ptr() in (
+            p.storage().data_ptr() for p in orig_module.parameters()
+        )
+
+    def _map_primal_to_param(
+        self, gm: fx.GraphModule, inputs: List[torch.Tensor], nparams: int
+    ) -> None:
+        self._dist_graph.primal_name_to_node.append({})
+        self._dist_graph.primal_to_param.append({})
+        for inp, node in zip(inputs, gm.graph.nodes):
+            if self._is_param(inp) and node.name.startswith("primals_"):
+                self._dist_graph.primal_name_to_node[0][node.name] = node
+                self._dist_graph.primal_to_param[0][node] = cast(
+                    nn.Parameter, inp
+                )
+        assert len(self._dist_graph.primal_name_to_node[0]) == nparams, (
+            len(self._dist_graph.primal_name_to_node[0]),
+            nparams,
+        )
+
+    def _map_grad_to_param(self, gm: fx.GraphModule) -> None:
+        # This assumes that the order of gradients are in the same order
+        # as the parameters in the input. The last argument of the output
+        # is None.
+        # TODO: this is very brittle, verify how we can do better than this.
+        self._dist_graph.grad_to_primal.append({})
```

Comment:
@fegin For my conceptual understanding, why is `grad_to_primal` a `List[Dict[fx.Node, fx.Node]]` and not just `Dict[fx.Node, fx.Node]`? The current implementation seems to assume that the list always has length 1 (e.g. L472).